src/rust/inetstack/protocols/layer4/tcp/established/sender.rs

```diff
@@ -121,35 +130,161 @@ impl Sender {
+        }
 
-        // The limited transmit algorithm can increase the effective size of cwnd by up to 2MSS.
-        let effective_cwnd: u32 = cwnd.get() + cb.congestion_control_get_limited_transmit_cwnd_increase().get();
+        self.unsent_next_seq_no = self.unsent_next_seq_no + (buf.len() as u32).into();
+        self.unsent_queue.push(Some(buf));
 
-        let win_sz: u32 = self.send_window.get();
+        // Wait until the sequnce number of the pushed buffer is acknowledged.
+        let mut send_unacked_watched: SharedAsyncValue<SeqNumber> = self.send_unacked.clone();
+        let ack_seq_no: SeqNumber = self.unsent_next_seq_no;
 
-        if win_sz > 0 && win_sz >= sent_data && effective_cwnd >= sent_data {
-            cmp::min(
-                cmp::min((win_sz - sent_data) as usize, self.mss),
-                (effective_cwnd - sent_data) as usize,
-            )
-        } else {
-            0
+        debug_assert!(send_unacked_watched.get() < ack_seq_no);
+        while send_unacked_watched.get() < ack_seq_no {
+            send_unacked_watched.wait_for_change(None).await?;
         }
+        Ok(())
     }
 
-    // Takes a segment and attempts to send it. Returns the number of bytes sent and any unsent data remains in
-    // `segment`.
-    pub fn send_segment(&mut self, segment: &mut DemiBuffer, cb: &mut SharedControlBlock) -> usize {
+    // Places a FIN marker in the outgoing data stream. No data can be pushed after this.
+    pub async fn push_fin_and_wait_for_ack(&mut self) -> Result<(), Fail> {
+        debug_assert!(self.fin_seq_no.is_none());
+        // TODO: We need to fix this the correct way: limit our send buffer size to the amount we're willing to buffer.
+        if self.unsent_queue.len() > UNSENT_QUEUE_CUTOFF {
+            return Err(Fail::new(EBUSY, "too many packets to send"));
+        }
+
+        self.fin_seq_no = Some(self.unsent_next_seq_no);
+        self.unsent_next_seq_no = self.unsent_next_seq_no + 1.into();
+        self.unsent_queue.push(None);
+
+        let mut send_unacked_watched: SharedAsyncValue<SeqNumber> = self.send_unacked.clone();
+        let fin_ack_num: SeqNumber = self.unsent_next_seq_no;
+        while self.send_unacked.get() < fin_ack_num {
+            send_unacked_watched.wait_for_change(None).await?;
+        }
+        Ok(())
+    }
+
+    pub async fn background_sender(&mut self, mut cb: SharedControlBlock) -> Result<Never, Fail> {
+        let mut send_unacked_watched: SharedAsyncValue<SeqNumber> = self.send_unacked.clone();
+        let mut cwnd_watched: SharedAsyncValue<u32> = cb.congestion_control_get_cwnd();
+
+        // The limited transmit algorithm may increase the effective size of cwnd by up to 2 * mss.
+        let mut ltci_watched: SharedAsyncValue<u32> = cb.congestion_control_get_limited_transmit_cwnd_increase();
+        let mut win_sz_watched: SharedAsyncValue<u32> = self.send_window.clone();
+
+        loop {
```

Comment:
This loop is very long - it's hard to keep track of the logic while reading top to bottom. Can it be broken into meaningful chunks? (e.g. the inner/nested loop and the if/else bodies)